### DISCUSSION TITLE
Max_scope is not working

### DISCUSSION THREAD

**Discussion by Gopalakrishnan11** [medspacy/medspacy#253] [General]
Hi,
In the context rule max scope is not working. I have mentioned max_scope is 5, but it's not limiting up to 5 tokens. It's going until it finds the target. Answers are welcome.
Thanks
![Screenshot from 2023-08-28 14-40-31](https://github.com/medspacy/medspacy/assets/105412331/722c4686-9e5d-47e8-9243-8ac8988ad1f5)

**Comment by abchapman93** [comment]
Hi @Gopalakrishnan11 ,
Could you provide some details of your code and output that I could review? I use max_scope often so it should be working, unless something broke in a recent update. What could be happening is that there's some overlapping or similar rule that doesn't have the max_scope argument set and that is actually the one modifying the entity.

For example, an example like this would be helpful:

```python
text = "" # The sentence you're showing in the example above
rule = ConTextRule(literal=..., category="NEGATED_EXISTENCE", max_scope=5) # The rule you expect to match

doc = nlp(text)
ent = doc.ents[...] # One of the entities being modified by the rule
ent._.modifiers[0].rule # Check that this is the same as the rule above 
```

**Reply by Gopalakrishnan11** [reply]
Hi @abchapman93 ,
Thanks for the response.
You can find the below function for the target rule and context rule. In the context rule for the negated term "denied", I mentioned max_scope is 5, but as you can see in my last discussion screenshot, it was not limiting up to 5 tokens.

If I check your code "ent._.modifiers[0].rule" the max scope is not showing
output :  ConTextRule(literal='deny', category='NEGATED_EXISTENCE', pattern=[{'LOWER': {'IN': ['deny', 'denied', 'denies', 'denying']}}], direction='FORWARD')

but I have added the max_scope in my json file.you can see the screenshot attached.
![Screenshot from 2023-08-28 20-45-10](https://github.com/medspacy/medspacy/assets/105412331/27ce397a-7ee3-43df-acba-6018e445330b)





def build_context():
        nlp = medspacy.load()
        with open('negpatterns.json') as f:
            context_rules = json.load(f)['context_rules']
        for rule in context_rules:
            context_rule = ConTextRule(
                literal=rule["literal"],
                category=rule["category"],
                direction=rule["direction"],
                max_scope=rule.get("max_scope",None),
                pattern=rule.get("pattern",None))
            nlp.get_pipe("medspacy_context").add(context_rule)
       with open('symptoms_target.json') as f:
            target_rules = json.load(f)['target_rule']
       for rule in target_rules:
            target_rule = TargetRule(
                rule["literal"],
                rule["category"],    
                rule.get("pattern", None))
               nlp.get_pipe("medspacy_target_matcher").add(target_rule)
        return nlp

**Comment by abchapman93** [comment]
I think that the screenshot cuts off some of the rules, is it the phrase "deny" that you're trying to limit the scope?

I think that the issue is that a rule included in the default ConText rules is superseding the new one that you're matching - i.e., they both match "denies", but the default rule which doesn't have `max_scope` is the one that is picked to create a modifier. To fix this, I recommend copying all of the default rules to a new file (you can find the [json file here](https://github.com/medspacy/medspacy/blob/master/resources/context_rules.json)) and then deleting the original "denies" rule. Then add these rules to your pipeline manually without loading the default rules.

For example:
```python
nlp = medspacy.load(disable=["medspacy_context"]) # Create an NLP pipeline without ConText
context = nlp.add_pipe("medspacy_context", config={"rules": None}) # Disable default rules
context.add(my_context_rules) # Add the rules you've developed
context.add(edited_default_rules) # Add the default rules from the json file with the one you want to remove deleted

```

Alternatively, you can also set a global max_scope option if you want to restrict the scope on _all_ rules:
```python
context = nlp.add_pipe("medspacy_context", config={"max_scope": 5})
```

Let me know if one of these solutions works for you.

**Reply by Gopalakrishnan11** [reply]
Hi @abchapman93,
I changed the json file and set the global max_scope, and now it's working.
Thanks for the response!

**Comment by abchapman93** [comment]
Great!
